File: fraiseql-kotlin/src/test/kotlin/com/fraiseql/ExportTypesTest.kt
package com.fraiseql

import com.fraiseql.schema.Schema
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.jsonObject
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.AfterEach
import org.junit.jupiter.api.Test
import java.io.File
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertNotNull
import kotlin.test.assertTrue

/**
 * Tests for minimal types.json export (TOML-based workflow)
 *
 * Validates that Schema.exportTypes() function generates minimal schema
 * with only types (no queries, mutations, observers, security, etc.)
 */
class ExportTypesTest {
    @BeforeEach
    fun setUp() {
        Schema.reset()
    }

    @AfterEach
    fun tearDown() {
        Schema.reset()
    }

    @Test
    fun testExportTypesMinimalSingleType() {
        // Register a single type
        Schema.registerType("User", mapOf(
            "id" to mapOf("type" to "ID", "nullable" to false),
            "name" to mapOf("type" to "String", "nullable" to false),
            "email" to mapOf("type" to "String", "nullable" to false),
        ), "User in the system")

        // Export minimal types
        val json = Schema.exportTypes(pretty = true)
        val parsed = Json.parseToJsonElement(json).jsonObject

        // Should have types section
        assertNotNull(parsed["types"])
        assertTrue(parsed["types"] is kotlinx.serialization.json.JsonArray)
        assertEquals(1, (parsed["types"] as kotlinx.serialization.json.JsonArray).size)

        // Should NOT have queries, mutations, observers
        assertFalse(parsed.containsKey("queries"))
        assertFalse(parsed.containsKey("mutations"))
        assertFalse(parsed.containsKey("observers"))
        assertFalse(parsed.containsKey("authz_policies"))

        // Verify User type
        val typesArray = parsed["types"] as kotlinx.serialization.json.JsonArray
        val userDef = typesArray[0].jsonObject
        assertEquals("User", userDef["name"]?.jsonPrimitive?.content)
        assertEquals("User in the system", userDef["description"]?.jsonPrimitive?.content)
    }

    @Test
    fun testExportTypesMultipleTypes() {
        // Register User type
        Schema.registerType("User", mapOf(
            "id" to mapOf("type" to "ID", "nullable" to false),
            "name" to mapOf("type" to "String", "nullable" to false),
        ))

        // Register Post type
        Schema.registerType("Post", mapOf(
            "id" to mapOf("type" to "ID", "nullable" to false),
            "title" to mapOf("type" to "String", "nullable" to false),
            "authorId" to mapOf("type" to "ID", "nullable" to false),
        ))

        // Export minimal
        val json = Schema.exportTypes(pretty = true)
        val parsed = Json.parseToJsonElement(json).jsonObject

        // Check types count
        val typesArray = parsed["types"] as kotlinx.serialization.json.JsonArray
        assertEquals(2, typesArray.size)

        // Verify both types present
        val typeNames = typesArray.map { it.jsonObject["name"]?.jsonPrimitive?.content }
        assertTrue(typeNames.contains("User"))
        assertTrue(typeNames.contains("Post"))
    }

    @Test
    fun testExportTypesNoQueries() {
        // Register type
        Schema.registerType("User", mapOf(
            "id" to mapOf("type" to "ID", "nullable" to false),
        ))

        // Export minimal
        val json = Schema.exportTypes(pretty = true)
        val parsed = Json.parseToJsonElement(json).jsonObject

        // Should have types
        assertNotNull(parsed["types"])

        // Should NOT have queries
        assertFalse(parsed.containsKey("queries"))
        assertFalse(parsed.containsKey("mutations"))
    }

    @Test
    fun testExportTypesCompactFormat() {
        // Register type
        Schema.registerType("User", mapOf(
            "id" to mapOf("type" to "ID", "nullable" to false),
        ))

        // Export compact (pretty=false)
        val compact = Schema.exportTypes(false)
        val pretty = Schema.exportTypes(true)

        // Both should be valid JSON
        assertNotNull(Json.parseToJsonElement(compact))
        assertNotNull(Json.parseToJsonElement(pretty))

        // Compact should be smaller or equal
        assertTrue(compact.length <= pretty.length)
    }

    @Test
    fun testExportTypesPrettyFormat() {
        // Register type
        Schema.registerType("User", mapOf(
            "id" to mapOf("type" to "ID", "nullable" to false),
        ))

        // Export pretty
        val json = Schema.exportTypes(true)

        // Should contain newlines (pretty format)
        assertTrue(json.contains("\n"))

        // Should be valid JSON
        assertNotNull(Json.parseToJsonElement(json))
    }

    @Test
    fun testExportTypesFile() {
        // Register type
        Schema.registerType("User", mapOf(
            "id" to mapOf("type" to "ID", "nullable" to false),
            "name" to mapOf("type" to "String", "nullable" to false),
        ))

        // Export to temporary file
        val tmpFile = "/tmp/fraiseql_types_test_kotlin.json"

        // Clean up if exists
        File(tmpFile).delete()

        // Export to file
        Schema.exportTypesFile(tmpFile)

        // Verify file exists
        assertTrue(File(tmpFile).exists())

        // Verify content
        val content = File(tmpFile).readText()
        val parsed = Json.parseToJsonElement(content).jsonObject

        assertNotNull(parsed["types"])
        val typesArray = parsed["types"] as kotlinx.serialization.json.JsonArray
        assertEquals(1, typesArray.size)

        // Cleanup
        File(tmpFile).delete()
    }

    @Test
    fun testExportTypesEmpty() {
        // Export with no types registered
        val json = Schema.exportTypes(true)
        val parsed = Json.parseToJsonElement(json).jsonObject

        // Should still have types key (as empty array)
        assertNotNull(parsed["types"])
        val typesArray = parsed["types"] as kotlinx.serialization.json.JsonArray
        assertEquals(0, typesArray.size)
    }
}

File: fraiseql-php/tests/ExportTypesTest.php
<?php

namespace FraiseQL\Tests;

use FraiseQL\Schema;
use FraiseQL\TypeInfo;
use FraiseQL\FieldDefinition;
use PHPUnit\Framework\TestCase;

/**
 * Tests for minimal types.json export (TOML-based workflow)
 *
 * Validates that ExportTypes() function generates minimal schema
 * with only types (no queries, mutations, observers, etc.)
 */
class ExportTypesTest extends TestCase
{
    protected function setUp(): void
    {
        Schema::reset();
    }

    protected function tearDown(): void
    {
        Schema::reset();
    }

    public function testExportTypesMinimalSingleType(): void
    {
        // Register a single type with fields
        Schema::registerType('User', [
            'name' => 'User',
            'description' => 'User in the system',
            'fields' => [
                ['name' => 'id', 'type' => 'ID', 'nullable' => false],
                ['name' => 'name', 'type' => 'String', 'nullable' => false],
                ['name' => 'email', 'type' => 'String', 'nullable' => false],
            ]
        ]);

        // Export minimal types
        $json = Schema::exportTypes(true);
        $parsed = json_decode($json, true);

        // Should have types section
        $this->assertArrayHasKey('types', $parsed);
        $this->assertIsArray($parsed['types']);
        $this->assertCount(1, $parsed['types']);

        // Should NOT have queries, mutations, observers
        $this->assertArrayNotHasKey('queries', $parsed);
        $this->assertArrayNotHasKey('mutations', $parsed);
        $this->assertArrayNotHasKey('observers', $parsed);
        $this->assertArrayNotHasKey('authz_policies', $parsed);

        // Verify User type
        $userDef = $parsed['types'][0];
        $this->assertEquals('User', $userDef['name']);
        $this->assertEquals('User in the system', $userDef['description']);
    }

    public function testExportTypesMultipleTypes(): void
    {
        // Register User type
        $userType = new TypeInfo('User', '');
        $userType->addField(new FieldDefinition('id', 'ID', false));
        $userType->addField(new FieldDefinition('name', 'String', false));
        Schema::registerType('User', $userType);

        // Register Post type
        $postType = new TypeInfo('Post', '');
        $postType->addField(new FieldDefinition('id', 'ID', false));
        $postType->addField(new FieldDefinition('title', 'String', false));
        $postType->addField(new FieldDefinition('authorId', 'ID', false));
        Schema::registerType('Post', $postType);

        // Export minimal
        $json = Schema::exportTypes(true);
        $parsed = json_decode($json, true);

        // Check types count
        $this->assertCount(2, $parsed['types']);

        // Verify both types present
        $typeNames = array_map(function ($t) { return $t['name']; }, $parsed['types']);
        $this->assertContains('User', $typeNames);
        $this->assertContains('Post', $typeNames);
    }

    public function testExportTypesNoQueries(): void
    {
        // Register type
        $userType = new TypeInfo('User', '');
        $userType->addField(new FieldDefinition('id', 'ID', false));
        Schema::registerType('User', $userType);

        // Export minimal
        $json = Schema::exportTypes(true);
        $parsed = json_decode($json, true);

        // Should have types
        $this->assertArrayHasKey('types', $parsed);

        // Should NOT have queries
        $this->assertArrayNotHasKey('queries', $parsed);
        $this->assertArrayNotHasKey('mutations', $parsed);
    }

    public function testExportTypesCompactFormat(): void
    {
        // Register type
        $userType = new TypeInfo('User', '');
        $userType->addField(new FieldDefinition('id', 'ID', false));
        Schema::registerType('User', $userType);

        // Export compact (pretty=false)
        $compactJson = Schema::exportTypes(false);
        $prettyJson = Schema::exportTypes(true);

        // Both should be valid JSON
        $this->assertIsArray(json_decode($compactJson, true));
        $this->assertIsArray(json_decode($prettyJson, true));

        // Compact should be smaller
        $this->assertLessThan(strlen($prettyJson), strlen($compactJson) + 100);
    }

    public function testExportTypesPrettyFormat(): void
    {
        // Register type
        $userType = new TypeInfo('User', '');
        $userType->addField(new FieldDefinition('id', 'ID', false));
        Schema::registerType('User', $userType);

        // Export pretty
        $json = Schema::exportTypes(true);

        // Should contain newlines (pretty format)
        $this->assertStringContainsString("\n", $json);

        // Should be valid JSON
        $parsed = json_decode($json, true);
        $this->assertArrayHasKey('types', $parsed);
    }

    public function testExportTypesFile(): void
    {
        // Register type
        $userType = new TypeInfo('User', '');
        $userType->addField(new FieldDefinition('id', 'ID', false));
        $userType->addField(new FieldDefinition('name', 'String', false));
        Schema::registerType('User', $userType);

        // Export to temporary file
        $tmpFile = '/tmp/fraiseql_types_test_php.json';

        // Clean up if exists
        if (file_exists($tmpFile)) {
            unlink($tmpFile);
        }

        // Export to file
        Schema::exportTypesFile($tmpFile);

        // Verify file exists
        $this->assertTrue(file_exists($tmpFile));

        // Verify content
        $content = file_get_contents($tmpFile);
        $parsed = json_decode($content, true);

        $this->assertArrayHasKey('types', $parsed);
        $this->assertCount(1, $parsed['types']);

        // Cleanup
        unlink($tmpFile);
    }

    public function testExportTypesEmpty(): void
    {
        // Export with no types registered
        $json = Schema::exportTypes(true);
        $parsed = json_decode($json, true);

        // Should still have types key (as empty array)
        $this->assertArrayHasKey('types', $parsed);
        $this->assertIsArray($parsed['types']);
        $this->assertEmpty($parsed['types']);
    }
}
